[BREAKING] CIRC-608: Upgrade circulation to item-storage 8.0

bohdan-suprun opened the pull request:
https://issues.folio.org/browse/CIRC-608: 
This story includes changes required by stories with breaking changes:

* MODINVSTOR-283 Restrict item states
* MODINVSTOR-332 Item: Change array copyNumbers to simple property copyNumber (breaking)
* MODINVSTOR-392 Make item.status.date read only field.
* MODINVSTOR-425 Make item.status required.

## Purpose
<!--
  Why are you making this change? There is nothing more important
  to provide to the reviewer and to future readers than the cause
  that gave rise to this pull request. Be careful to avoid circular
  statements like "the purpose is to update the schema." and
  instead provide an explanation like "there is more data to be provided and stored for Purchase Orders 
  which is currently missing in the schema"

  The purpose may seem self-evident to you now, but the standard to
  hold yourself to should be "can a developer parachuting into this
  project reconstruct the necessary context merely by reading this
  section."

  If you have a relevant JIRA issue, add a link directly to the issue URL here.
  Example: https://issues.folio.org/browse/MODORDERS-70
 -->

## Approach
<!--
 How does this change fulfill the purpose? It's best to talk
 high-level strategy and avoid code-splaining the commit history.

 The goal is not only to explain what you did, but help other
 developers *work* with your solution in the future.
-->

#### TODOS and Open Questions
<!-- OPTIONAL
- [ ] Use GitHub checklists. When solved, check the box and explain the answer.
-->

## Learning
<!-- OPTIONAL
  Help out not only your reviewer, but also your fellow developer!
  Sometimes there are key pieces of information that you used to come up
  with your solution. Don't let all that hard work go to waste! A
  pull request is a *perfect opportunity to share the learning that
  you did. Add links to blog posts, patterns, libraries or addons used
  to solve this problem.
-->

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [x] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

id-jenkins commented:
<html><body><pre>
INFO: lint-raml: Investigating and determining configuration: mod-circulation/ramls
INFO: lint-raml: Not assessing schema descriptions, as per option '--validate-only'.
INFO: lint-raml: Assessing RAML files (https://dev.folio.org/guides/raml-cop/):
INFO: lint-raml-cop: Processing RAML v1.0 file: circulation.raml
ERROR: lint-raml-cop:   raml-cop detected errors with circulation.raml:
[ramls/examples/request.json:9:2] ERROR Content is not valid according to schema: Missing required property: status

INFO: lint-raml-cop: Processing RAML v1.0 file: end-patron-action-session.raml
INFO: lint-raml-cop:   raml-cop did not detect any errors with end-patron-action-session.raml
INFO: lint-raml-cop: Processing RAML v1.0 file: inventory-reports.raml
INFO: lint-raml-cop:   raml-cop did not detect any errors with inventory-reports.raml
INFO: lint-raml-cop: Processing RAML v1.0 file: loan-anonymization.raml
INFO: lint-raml-cop:   raml-cop did not detect any errors with loan-anonymization.raml
INFO: lint-raml-cop: Processing RAML v1.0 file: pick-slips.raml
INFO: lint-raml-cop:   raml-cop did not detect any errors with pick-slips.raml
INFO: lint-raml-cop: Processing RAML v1.0 file: request-move.raml
ERROR: lint-raml-cop:   raml-cop detected errors with request-move.raml:
[ramls/examples/request.json:9:2] ERROR Content is not valid according to schema: Missing required property: status

INFO: lint-raml-cop: Processing RAML v1.0 file: requests-reports.raml
INFO: lint-raml-cop:   raml-cop did not detect any errors with requests-reports.raml
ERROR: lint-raml: There were processing errors. See list above.
</pre><body></html>


Note: When those errors are fixed, then the INFO messages will be gone too. See Jenkins "Artifacts" tab.

marcjohnson-kint commented on ramls/check-in-by-barcode-response.json:
```diff
@@ -77,10 +77,18 @@
           },
-          "additionalProperties": false
+          "additionalProperties": false,
+          "required": ["name"]
```
What does a required property mean in a schema used only for responses, that it is guaranteed to be there?

bohdan-suprun replied:
Yes, I think so. Just for documentation purpose.

marcjohnson-kint replied:
This would mean that not including the name of the status was a breaking compatibility change?

What does the client gain from knowing this is required?

bohdan-suprun replied:
My main thought was that since Item.status will always be present, then it always be included in the response.

marcjohnson-kint replied:
> My main thought was that since Item.status will always be present, then it always be included in the response.

I think that starts to couple the public `circulation` API with expectations from the `item-storage` API.

If this is only for documentation purposes, I think we should consider deferring this change until we better understand the impact on maintainability of designs like this.

marcjohnson-kint commented on ramls/check-in-by-barcode-response.json:
```diff
@@ -77,10 +77,18 @@
               "description": "Name of the item status",
-              "type": "string"
+              "type": "string",
+              "$ref": "schema/item-status-names.json"
```
What is the purpose of an enumeration here?

I think we need to be really cautious with this approach. What happens when a new item status is introduced in inventory? Can that new status be presented here?

bohdan-suprun replied:
Well, this enum does not restrict anything, in a fact (because CIRC is not an RMB module, so there is no validation against schema). 

However, there was restriction before those changes. We've being used `ItemStatus.java` enum for item statuses, with defined elements.

marcjohnson-kint replied:
> Well, this enum does not restrict anything, in a fact (because CIRC is not an RMB module, so there is no validation against schema).

Whilst no representation is validated against this schema (yet), this schema forms part of a contract with the client.

If inventory introduces the `Recently returned` status and an item with that status is included in a check in response, what should the client do, given that the response is invalid according to the schema?

(RAML Module Builder also does not validate against schema for outgoing representations, only incoming).
 
> However, there was restriction before those changes. We've being used `ItemStatus.java` enum for item statuses, with defined elements.

That restriction is internal to the module, it is the server's choice that it only uses a certain set of statuses. The client is ignorant of this.

bohdan-suprun replied:
> If inventory introduces the Recently returned status and an item with that status is included in a check in response, what should the client do, given that the response is invalid according to the schema?

I expect that endpoint will require updates for this case in order to handle this item at all. We use the `ItemStatus` enum which do not contain this new value, and because of this the item won't be correctly processed I think.

marcjohnson-kint replied:
> > If inventory introduces the Recently returned status and an item with that status is included in a check in response, what should the client do, given that the response is invalid according to the schema?
> 
> I expect that endpoint will require updates for this case in order to handle this item at all. 

You expect that the `circulation` interface has to change every time the `item-storage` interface changes the list of statuses?

> We use the `ItemStatus` enum which do not contain this new value, and because of this the item won't be correctly processed I think.

Does check in fail if an item is in an unknown status prior to check in? Do all circulation operations fail if the status is not understood?

bohdan-suprun replied:
I've double checked the logic and it will be possible to make check in for an item with a status that are not present in circulation's enum, because check in operation always puts the item into available/in transit/awaiting pickup/delivery.

I'm not sure about another circulation operations.

marcjohnson-kint commented on descriptors/ModuleDescriptor-template.json:
```diff
@@ -108,7 +108,7 @@
     {
       "id": "circulation",
-      "version": "8.5",
+      "version": "9.0",
```
The implementation version also needs to change.

Does this mean that a lot more coordinated pull requests are needed before any of these changes can be merged?

bohdan-suprun replied:
Yes, mostly because of making item.copyNumbers to be a simple property. It was exposed in requests API.

bohdan-suprun replied:
The implementation version has been changed.

marcjohnson-kint commented on ramls/loan.json:
```diff
@@ -134,9 +134,18 @@
+              "type": "string",
+              "$ref": "schema/item-status-names.json"
+            },
+            "date": {
```
Why is this property being introduced?

bohdan-suprun replied:
It was just missed in schema definition, but it is being added in ItemSummaryRepresentation.

marcjohnson-kint replied:
Why was it included in the item representation? Was this deliberately done for all situations where that class was used, and therefore the bug is with the interface definition? Or should it not be included in all of those cases, and the bug is with the implementation?

bohdan-suprun replied:
It is related to https://issues.folio.org/browse/CIRC-523. @VRohach, do we need `status.date` in the loan response?

VRohach replied:
@bohdan-suprun In the current implementation, the item.status.date returning everywhere, where item uses. So, in that cases when item exists in the loan, item.status.date will be there (if it isn`t empty).

bohdan-suprun replied:
Thanks for clarification, @VRohach. Will leave the declaration in json schema then.

marcjohnson-kint commented on ramls/request.json:
```diff
@@ -74,6 +74,12 @@
       "readOnly": true,
       "type": "object",
       "properties": {
+        "status": {
```
Is this a new property? If so, why is it being introduced?

bohdan-suprun replied:
Same as above, it was missed in the schema definition, see [RequestRepresentation.java#L87](https://github.com/folio-org/mod-circulation/blob/master/src/main/java/org/folio/circulation/domain/RequestRepresentation.java#L87)

marcjohnson-kint replied:
Similarly, was it introduced intentionally and the bug was in the interface definition or accidentally because of shared representation classes and the bug is in the implementation?

bohdan-suprun replied:
I'm not sure, the property was added almost 2 years ago. It seems that it is an interface issue.

marcjohnson-kint replied:
Ok, I'll close resolve

marcjohnson-kint commented on src/main/java/org/folio/circulation/domain/ItemStatus.java:
```diff
@@ -10,11 +10,12 @@
   ON_ORDER("On order"),
-  IN_PROCESS("In process"),
+  PAGED("Paged"),
+  WITHDRAWN("Withdrawn"),
```
Have these new states been included in the request white / black list? And in rules for checking out?

bohdan-suprun replied:
No, I can remove these new statuses, they were added just to be consistent with the enum values in the schema.

Does it work?

marcjohnson-kint replied:
I think we need to find out what circulation should do with these statuses, please can you follow up with Emma and Cate.

I'm slightly surprised `Paged` wasn't present, as that is a status that circulation actively uses.

bohdan-suprun replied:
Sorry, it seems that values are reordered. The only new status here is `WITHDRAWN`.

marcjohnson-kint replied:
What does withdrawn mean?

bohdan-suprun replied:
I don't know, maybe it is an order-related status. But this status is present on the inventory.item search tab:
![image](https://user-images.githubusercontent.com/9094650/71823542-2280fa00-30a0-11ea-803f-e445045fd96f.png)

bohdan-suprun replied:
I don't know, maybe it is an order-related status. But this status is present on the inventory.item search tab:
![image](https://user-images.githubusercontent.com/9094650/71823542-2280fa00-30a0-11ea-803f-e445045fd96f.png)

marcjohnson-kint replied:
Please can you follow up with Cate, Charlotte and Emma to try to identify why it was introduced and how circulation should handle that status.

bohdan-suprun replied:
I'll move it as a separate ticket.

bohdan-suprun replied:
Removed as per clarification.

marcjohnson-kint commented on src/test/java/api/item/ItemLastCheckInTests.java:
```diff
@@ -77,9 +77,10 @@ public void shouldNotFailCheckInWithInvalidLoggedInUserId() {
     final OkapiHeaders okapiHeaders = getOkapiHeadersFromContext()
       .withRequestId("check-in-by-barcode-request")
-      .withUserId("INVALID_UUID");
+      .withUserId(invalidUuid);
```
Why did this need to change?

bohdan-suprun replied:
I've introduced validation of request against schema for fake item-storage, and `INVALID_UUID` was not a valid value.

marcjohnson-kint replied:
Does this invalidate the test, as an invalid user ID, that isn't a UUID, will cause this to fail?

bohdan-suprun replied:
I think the purpose of this test to verify that checkin won't be failed when there is no user with such ID.

marcjohnson-kint replied:
My understanding of `invalid` is that the ID provided is in an invalid format, not that the ID does not reference a known record, and hence the user is unknown.

I think that is why an invalid UUID was used, rather than a randomly generated UUID that does not refer to an existing user.

bohdan-suprun replied:
In this case, the test case is not relevant. Because item-storage is not allowing such values.

bohdan-suprun replied:
Hi @mrdpopov, it seems that this is your test, could you please clarify what is being verified by the case? Are we verifying that check-in works even if the user not found or the user UUID has an invalid format?

mrdpopov replied:
The purpose of the test is to show that the last checkin does not fail when the supplied user id is invalid.  
I think it is relevant because It shows that passed userID is stored stored in the last checkin object as is(e.g case when userId is not found). I believe, this was in the requirements.

bohdan-suprun replied:
Ok, will resolve the comment in this case.

marcjohnson-kint commented on src/test/java/api/requests/RequestsAPICreationTests.java:
```diff
@@ -378,7 +377,7 @@ public void canCreateARequestEvenWithDifferentCaseCheckedOutStatus() {
 
     JsonObject itemWithChangedStatus = smallAngryPlanet.copyJson()
-      .put("status", new JsonObject().put("name", "Checked Out"));
+      .put("status", new JsonObject().put("name", "Checked out"));
```
Why has this been changed? Is this still possible to do?

bohdan-suprun replied:
You're right, it is not possible to do anymore. Because `inventory` does not allow such status.

marcjohnson-kint commented on src/test/java/api/support/builders/ItemBuilder.java:
```diff
@@ -34,14 +35,14 @@
   public ItemBuilder() {
     this(UUID.randomUUID(), null, "565578437802", AVAILABLE,
-      null, null, null, null, null, null, null, null, null, null, null, null,
+      null, null, null, null, null, null, null, null, null, null, null, Collections.emptyList(),
```
What parameter is an empty collection being passed to?

bohdan-suprun replied:
It is `yearCaption` which is an array.

marcjohnson-kint replied:
Ok, thanks

marcjohnson-kint commented on src/test/java/api/support/fakes/StorageSchema.java:
```diff
@@ -7,6 +7,10 @@
   private StorageSchema() { }
 
+  public static JsonSchemaValidator validatorForStorageItemSchema() throws IOException {
+    return JsonSchemaValidator.fromResource("/item-storage-8-0.json");
```
I’m glad this was introduced, thank you

marcjohnson-kint commented on src/test/java/api/support/fixtures/TemplateContextMatchers.java:
```diff
@@ -66,7 +64,7 @@ private TemplateContextMatchers() {
     tokenMatchers.put("item.callNumberPrefix", is(callNumberPrefix));
     tokenMatchers.put("item.callNumberSuffix", is(callNumberSuffix));
-    tokenMatchers.put("item.copy", is(copyNumbers));
+    tokenMatchers.put("item.copy", is(copyNumber));
```
Does this impact the template definitions for patron notices?

bohdan-suprun replied:
No, when we had several values for `item.copyNumbers` they were built as a single string, separated by a semicolon.

I think it will work fine for a single value as well.

bohdan-suprun commented:
Hi @marcjohnson-kint, I've removed the status validation with enum, because it used in responses only and it won't be so useful but introduces coupling with inventory instead.

marcjohnson-kint commented on ramls/check-in-by-barcode-response.json:
```diff
@@ -78,6 +78,11 @@
+            "date": {
+              "type": "string",
+              "description": "Date time when status was last changed",
+              "format": "date-time"
```
Was this already being provided, and missing in the interface definition, or is it a new property?

bohdan-suprun replied:
It is an existing property. It is being set in [ItemSummaryRepresentation.java#L44](https://github.com/folio-org/mod-circulation/blob/master/src/main/java/org/folio/circulation/domain/representations/ItemSummaryRepresentation.java#L44) which called in `org.folio.circulation.domain.representations.CheckInByBarcodeResponse#mapToResponse`

marcjohnson-kint commented on src/test/java/api/requests/RequestsAPICreationTests.java:
```diff
@@ -1559,7 +1528,7 @@ public void recallRequestNoticeIsSentWhenPolicyDefinesRecallRequestNoticeConfigu
       "ItemPrefix",
       "ItemSuffix",
-      Arrays.asList("CopyNumbers", "CopyDetails"));
+      "CopyNumbers");
```
There can only be one copy number now. Maybe this string should be changed to reflect that?

marcjohnson-kint commented on src/test/java/api/support/fakes/FakeOkapi.java:
```diff
@@ -5,6 +5,8 @@
 import static api.support.fakes.StorageSchema.validatorForLocationInstSchema;
 import static api.support.fakes.StorageSchema.validatorForLocationLibSchema;
+import static api.support.fakes.StorageSchema.validatorForStorageItemSchema;
+import static api.support.fakes.StorageSchema.validatorForStorageItemSchema;
```
Why is this import included twice?

marcjohnson-kint commented on ramls/request.json:
```diff
@@ -135,12 +140,9 @@
-          "items": {
-            "type": "string"
-          }
+        "copyNumber": {
```
Should this be read-only?

bohdan-suprun replied:
Following properties have been declared as read only:
* item.* (already readonly, only few properties was missed)
* requester.* (already readonly, only few properties was missed)
* proxy.* (already readonly, only few properties was missed)
* deliveryAddress.*
* pickupServicePoint.*

marcjohnson-kint replied:
Thank you

marcjohnson-kint commented on ramls/loan.json:
```diff
@@ -135,8 +135,15 @@
             "name": {
```
Given we are making a breaking change, should this be read-only as well?

Please can you check for other properties that should be read-only and aren't marked as such.

bohdan-suprun replied:
Sure. Following properties have been declared as read only:
* item.*
* checkoutServicePoint.*
* checkinServicePoint.*

marcjohnson-kint replied:
Thank you

marcjohnson-kint commented:
@bohdan-suprun Given that compatibility is being broken, are there any more minor improvements that can be made to the schema or RAML.

For example, can we safely add UUID and date-time formats to properties?

bohdan-suprun commented:
> @bohdan-suprun Given that compatibility is being broken, are there any more minor improvements that can be made to the schema or RAML.
> 
> For example, can we safely add UUID and date-time formats to properties?

Let me check. The date-time format should be confirmed with the storage modules I think, because my understanding that `date-time` format means the RMB defined standard of timestamps, or is it fine to set it for fields that intended to be date-time?

marcjohnson-kint commented:
> > @bohdan-suprun Given that compatibility is being broken, are there any more minor improvements that can be made to the schema or RAML.
> > For example, can we safely add UUID and date-time formats to properties?
> 
> The date-time format should be confirmed with the storage modules I think

What do you mean by confirmed?

> because my understanding that `date-time` format means the RMB defined standard of timestamps

The [date-time](https://json-schema.org/draft/2019-09/json-schema-validation.html#rfc.section.7.3.1) format is part of the JSON Schema specification and is based upon RFC339, it is not specific to RAML Module Builder.

(Based upon the example you gave me, RAML Module Builder produces date and times that *are not* compliant with RFC-3339, most of the usages likely work because they are based upon the broader ISO8601 standard)

> or is it fine to set it for fields that intended to be date-time?

I think that is what we should do, as the API for mod-circulation should be self contained where possible and not be dependent upon the storage APIs, even implicitly.

bohdan-suprun commented:
Ok, that make sense @marcjohnson-kint, will investigate what we can improve.

id-jenkins commented:
<html><body><pre>
INFO: lint-raml: Investigating and determining configuration: mod-circulation/ramls
WARNING: lint-raml: Missing from configuration: circulation-rules.raml
WARNING: lint-raml: Configuration needs to be updated (FOLIO-903).
INFO: lint-raml: Not assessing schema descriptions, as per option '--validate-only'.
INFO: lint-raml: Assessing RAML files (https://dev.folio.org/guides/raml-cop/):
INFO: lint-raml-cop: Processing RAML v1.0 file: circulation-rules.raml
INFO: lint-raml-cop:   raml-cop did not detect any errors with circulation-rules.raml
INFO: lint-raml-cop: Processing RAML v1.0 file: circulation.raml
INFO: lint-raml-cop:   raml-cop did not detect any errors with circulation.raml
INFO: lint-raml-cop: Processing RAML v1.0 file: declare-item-lost.raml
INFO: lint-raml-cop:   raml-cop did not detect any errors with declare-item-lost.raml
INFO: lint-raml-cop: Processing RAML v1.0 file: end-patron-action-session.raml
INFO: lint-raml-cop:   raml-cop did not detect any errors with end-patron-action-session.raml
INFO: lint-raml-cop: Processing RAML v1.0 file: inventory-reports.raml
INFO: lint-raml-cop:   raml-cop did not detect any errors with inventory-reports.raml
INFO: lint-raml-cop: Processing RAML v1.0 file: loan-anonymization.raml
INFO: lint-raml-cop:   raml-cop did not detect any errors with loan-anonymization.raml
INFO: lint-raml-cop: Processing RAML v1.0 file: pick-slips.raml
INFO: lint-raml-cop:   raml-cop did not detect any errors with pick-slips.raml
INFO: lint-raml-cop: Processing RAML v1.0 file: request-move.raml
ERROR: lint-raml-cop:   raml-cop detected errors with request-move.raml:
[ramls/examples/request-move.json:0:0] ERROR Content is not valid according to schema: Missing required property: id

INFO: lint-raml-cop: Processing RAML v1.0 file: requests-reports.raml
INFO: lint-raml-cop:   raml-cop did not detect any errors with requests-reports.raml
ERROR: lint-raml: There were processing errors. See list above.
</pre><body></html>


Note: When those errors are fixed, then the INFO messages will be gone too. See Jenkins "Artifacts" tab.

bohdan-suprun commented:
@marcjohnson-kint, I've added UUID patterns for loan and request schemes, also added minimum restriction for request.position (1) and loan.renewalCount (0).  

request-move schema did not have id property (there were only destinationItemId and requestType).

marcjohnson-kint commented:
> @marcjohnson-kint, I've added UUID patterns for loan and request schemes, also added minimum restriction for request.position (1) and loan.renewalCount (0).

Thank you.

> request-move schema did not have id property (there were only destinationItemId and requestType).

I think there might be some confusion around this API. The implementation uses a path parameter for the ID of the request rather than a property in the request body. It appears the builder the test uses does create an `id` property which seems incorrect to me.

I don't think a property in the body should be needed, and should be removed from being used in the tests.

marcjohnson-kint commented on ramls/request-move.json:
```diff
@@ -3,9 +3,15 @@
+      "description": "Id of request that have to be moved",
+      "type": "string",
+      "$ref": "uuid.json"
+    },
```
I believe the `id` of the request comes from a template parameter in the path, not from the body of the request (see my general comment).

bohdan-suprun replied:
Aha, thanks, have not seen the parameter. Will remove it.

marcjohnson-kint commented:
@bohdan-suprun  Please update the example request move request JSON, as it is now invalid according to the schema

bohdan-suprun commented:
> @bohdan-suprun Please update the example request move request JSON, as it is now invalid according to the schema

Hi @marcjohnson-kint, it should be valid https://github.com/folio-org/mod-circulation/blob/CIRC-608-item-storage-8/ramls/examples/request-move.json.

I've reverted it in this commit c520e10 (force-pushed).
